src/rest/resource.cpp

```diff
@@ -499,75 +505,179 @@ void Resource::Rloc(const Request &aRequest, Response &aResponse) const
     }
 }
 
-void Resource::GetActiveDatasetTlvs(Response &aResponse) const
+void Resource::GetDataset(DatasetType aDatasetType, const Request &aRequest, Response &aResponse) const
 {
-    otOperationalDatasetTlvs datasetTlvs;
-    otError                  error = OT_ERROR_NONE;
+    otbrError                error = OTBR_ERROR_NONE;
+    struct NodeInfo          node;
     std::string              body;
     std::string              errorCode;
+    otOperationalDataset     dataset;
+    otOperationalDatasetTlvs datasetTlvs;
 
-    SuccessOrExit(error = otDatasetGetActiveTlvs(mInstance, &datasetTlvs));
+    if (aRequest.GetHeaderValue(OT_REST_ACCEPT_HEADER) == OT_REST_CONTENT_TYPE_PLAIN)
+    {
+        if (aDatasetType == DatasetType::kActive)
+        {
+            VerifyOrExit(otDatasetGetActiveTlvs(mInstance, &datasetTlvs) == OT_ERROR_NONE,
+                         error = OTBR_ERROR_NOT_FOUND);
+        }
+        else if (aDatasetType == DatasetType::kPending)
+        {
+            VerifyOrExit(otDatasetGetPendingTlvs(mInstance, &datasetTlvs) == OT_ERROR_NONE,
+                         error = OTBR_ERROR_NOT_FOUND);
+        }
 
-    body = Json::Bytes2HexJsonString(datasetTlvs.mTlvs, datasetTlvs.mLength);
+        aResponse.SetContentType(OT_REST_CONTENT_TYPE_PLAIN);
+        body = Json::Bytes2HexString(datasetTlvs.mTlvs, datasetTlvs.mLength);
+    }
+    else
+    {
+        if (aDatasetType == DatasetType::kActive)
+        {
+            VerifyOrExit(otDatasetGetActive(mInstance, &dataset) == OT_ERROR_NONE, error = OTBR_ERROR_NOT_FOUND);
+        }
+        else if (aDatasetType == DatasetType::kPending)
+        {
+            VerifyOrExit(otDatasetGetPending(mInstance, &dataset) == OT_ERROR_NONE, error = OTBR_ERROR_NOT_FOUND);
+        }
+        body = Json::Dataset2JsonString(dataset);
+    }
 
     aResponse.SetBody(body);
-    errorCode = GetHttpStatus(HttpStatusCode::kStatusOk);
-    aResponse.SetResponsCode(errorCode);
 
 exit:
-    if (error != OT_ERROR_NONE)
+    if (error == OTBR_ERROR_NONE)
+    {
+        errorCode = GetHttpStatus(HttpStatusCode::kStatusOk);
+        aResponse.SetResponsCode(errorCode);
+    }
+    else if (error == OTBR_ERROR_NOT_FOUND)
+    {
+        errorCode = GetHttpStatus(HttpStatusCode::kNoContent);
+        aResponse.SetResponsCode(errorCode);
+    }
```

Comment:
Try adding a `SetResponsCode(otbrError aError)` method to `Response` so that we can handle common `otbrError` to Response status code conversion in a single place.

Reply:
Doing the response status code conversion in `Response` seems the right place indeed, as this is a HTTP thing really. However, we do use it in `void Resource::ErrorHandler(Response &aResponse, HttpStatusCode aErrorCode) const` as well, which generates a JSON body with the same string. This is more an implementation specific handling of the HTTP body, which probably should stay in `Resource`?

If we look at `Response` as an "OTBR opinionated" Response implementation, then maybe we can move ErrorHandler to `Response` as well (e.g. `SetErrorResponse(HttpStatusCode aCode)`?)

It kinda goes into refactoring real which I'd prefer to do in a follow up.